pkg/charts/defaults.go

```diff
@@ -0,0 +1,265 @@
+package charts
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/helm/pkg/chartutil"
+)
+
+const (
+	helmDefaultValuesFile   = "values.yaml"
+	helmDefaultHAValuesFile = "values-ha.yaml"
+)
+
+// DefaultValues contain all the default variables defined in the values.yaml
+// and values-ha.yaml.
+type DefaultValues struct {
```

Comment:
It seems there is overlap between this struct, `charts.Values` and `installValues` used in `install.go`. The other structs in `charts.Values` already reflect the contents of `values.yaml` (Proxy, init, ports, etc). What do you think about completing `charts.Values` so it completely reflects `values.yaml` and use that everywhere we need to pass values to Helm? This wouldn't affect the injection side of things as that only uses the stuff in `charts.Values`  it needs.

Reply:
The thought did cross my mind, but I decided against it for now, mainly because I don't want to deal with multiple layers of type casting. The default values read from the `values*.yaml` will be of primitive types. In order to get them into `charts.Values`, I'll have to convert them into custom types defined in the `pkg/charts` package, only to have them reconverted back to primitive types to be rendered in the templates. I feel we might end up adding more code, than saving.

IMO, it's also reasonable to keep `charts.Values` as an inject-only struct. The context it is used (by its surrounding code) is different from that of the default values. I think at one point, we chatted about whether to name it as `charts.InjectValues`.

If we feel like it's a worthwhile refactoring effort, perhaps we can track it in a separate PR.

Reply:
I think your intuition is right. I'll like to give this some more thought post-2.5, as the amount of code change will be non-trivial (with the removal of `installValues`). I've created https://github.com/linkerd/linkerd2/issues/3237 to track the refactor work.

Reply:
Maybe I'm misunderstanding the problem, but take this for example:
```go
for _, valuesFile := range valuesFiles {
		valuesStruct := &Values{}
		if err := yaml.Unmarshal(valuesFile.Data, valuesStruct); err != nil {
			fmt.Printf("error: %s\n", err)
		}
		fmt.Printf("valuesStruct: %#v\n", valuesStruct)
		fmt.Printf("valuesStruct.Issuer.ClockSkewAllowance: %s\n", valuesStruct.Identity.Issuer.ClockSkewAllowance)
		v, err := chartutil.ReadValues(valuesFile.Data)
		if err != nil {
			return nil, err
		}
		values.MergeInto(v)
	}
```
Which outputs:
```
valuesStruct: &charts.Values{Namespace:"linkerd", ClusterDomain:"cluster.local", HighAvailability:false, Identity:(*charts.Identity)(0xc00056d830), Proxy:(*charts.Proxy)(0xc00004d740), ProxyInit:(*charts.ProxyInit)(0xc0002f7780)}
valuesStruct.Issuer.ClockSkewAllowance: 20s
```
So we could replace the call `chartutil.ReadValues(valuesFile.Data)` and do the unmarshalling ourselves with  `yaml.Unmarshal(valuesFile.Data, valuesStruct)` which gives us a well-formed `charts.Values` (pending adding the missing fields), instead of having to deal with the flat `chartutil.Values` map...